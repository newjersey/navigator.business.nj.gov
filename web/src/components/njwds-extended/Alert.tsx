import React, { ReactElement, ReactNode } from "react";

interface Props {
  variant: AlertVariant;
  children: ReactNode;
  noIcon?: boolean;
  heading?: string;
  rounded?: boolean;
  dataTestid?: string;
  className?: string;
}

export type AlertVariant = "info" | "success" | "warning" | "error";

export const Alert = (props: Props): ReactElement => {
  const { variant, children, noIcon, heading, rounded, dataTestid } = props;
  const variantClass = variant ? `usa-alert--${variant}` : "";
  const noIconClass = noIcon ? " usa-alert--no-icon" : "";
  const roundedClass = rounded ? " radius-md" : "";
  const alertRole = variant === "error" ? { role: "alert" } : {};
  const defaultClassNames = "usa-alert margin-y-2 usa-alert--slim";
  const className = [defaultClassNames, roundedClass, variantClass, noIconClass, props.className ?? ""]
    .map((i) => i?.trim())
    .filter((value: string | undefined) => value && value.length > 0)
    .join(" ");

  return (
    <div className={className} {...alertRole} {...(dataTestid ? { "data-testid": dataTestid } : {})}>
      <div className="usa-alert__body">
        {heading && <h3 className="margin-bottom-0">{heading}</h3>}
        <div className="usa-alert__text">{children}</div>
      </div>
    </div>
  );
};
