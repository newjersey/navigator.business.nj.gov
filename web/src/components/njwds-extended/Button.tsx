import { CircularProgress } from "@mui/material";
import React, { ReactElement, useEffect, useRef, useState } from "react";
interface Props {
  style:
    | "primary"
    | "secondary"
    | "tertiary"
    | "primary-big"
    | "secondary-big"
    | "secondary-input-field-height"
    | "primary-input-field-height";
  children: React.ReactNode;
  onClick?: (() => void) | ((event: React.MouseEvent) => Promise<void>) | ((event: React.MouseEvent) => void);
  dataTestid?: string;
  typeSubmit?: boolean;
  noRightMargin?: boolean;
  underline?: boolean;
  smallText?: boolean;
  textBold?: boolean;
  loading?: boolean;
  widthAutoOnMobile?: boolean;
  heightAutoOnMobile?: boolean;
}

export const Button = (props: Props): ReactElement => {
  let style = "";
  let disabledClass = "";
  const widthRef = useRef<HTMLInputElement | null>(null);
  const [width, setWidth] = useState<number>();
  const [height, setHeight] = useState<number>();

  switch (props.style) {
    case "primary":
      style = "usa-button padding-y-1";
      disabledClass = "usa-button--disabled";
      break;
    case "primary-input-field-height":
      style = "usa-button padding-y-2";
      disabledClass = "usa-button--disabled padding-y-2";
      break;
    case "secondary":
      style = "usa-button usa-button--outline padding-y-1";
      disabledClass = "usa-button--outline-disabled";
      break;
    case "secondary-input-field-height":
      style = "usa-button usa-button--outline padding-y-2";
      disabledClass = "usa-button--outline-disabled padding-y-2";
      break;
    case "tertiary":
      style = "usa-button usa-button--unstyled";
      break;
  }
  const showDisabledClass = props.loading ? disabledClass : "";

  useEffect(() => {
    if (!props.loading && widthRef.current !== null) {
      setWidth(widthRef.current.clientWidth);
      setHeight(widthRef.current.clientHeight);
    }
  }, [height, width, props.loading, disabledClass]);

  return (
    <button
      className={`${style} ${props.heightAutoOnMobile ? "height-auto" : ""}${
        props.noRightMargin && props.style !== "tertiary" ? " margin-right-0" : ""
      }${!props.noRightMargin && props.style !== "tertiary" ? " margin-right-2" : ""}${
        props.style === "tertiary" ? " margin-right-0" : ""
      }${props.underline ? " underline" : ""}${props.smallText ? " font-body-2xs" : ""}${
        props.textBold ? " text-bold" : ""
      }${props.widthAutoOnMobile ? " width-auto" : ""} ${showDisabledClass}`}
      onClick={props.onClick}
      {...(props.typeSubmit ? { type: "submit" } : { type: "button" })}
      {...(props.dataTestid ? { "data-testid": props.dataTestid } : {})}
    >
      {props.loading ? (
        <div style={{ width: width, height: height }} data-testid="loading-spinner">
          <div className="margin-top-neg-1">
            <CircularProgress size={24} thickness={10} />
          </div>
        </div>
      ) : (
        <div ref={widthRef}>{props.children}</div>
      )}
    </button>
  );
};
