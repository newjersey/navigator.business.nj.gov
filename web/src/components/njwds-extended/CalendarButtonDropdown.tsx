import { UnStyledButton } from "@/components/njwds-extended/UnStyledButton";
import { Icon } from "@/components/njwds/Icon";
import { Box, ClickAwayListener, Grow, MenuItem, MenuList, Paper, Popper } from "@mui/material";
import React, { ReactElement, ReactNode, useEffect, useRef, useState, type JSX } from "react";

type ButtonDropdownOption = {
  text: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  props?: Record<string, any>;
  onClick: () => void;
};

interface Props {
  children: ReactNode;
  name?: string;
  dropdownOptions: ButtonDropdownOption[];
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  wrapper?: (props: { children: ReactNode; [key: string]: any }) => ReactElement<any>;
  dropdownClassName?: string;
  horizontal?: boolean;
  hideDivider?: boolean;
}

export const CalendarButtonDropdown = (props: Props): ReactElement<any> => {
  const [open, setOpen] = useState(false);
  const anchorRef = useRef<HTMLButtonElement | null>(null);
  const prevOpen = useRef(open);
  useEffect(() => {
    if (null !== anchorRef.current && prevOpen.current && !open) {
      anchorRef.current.focus();
    }
    prevOpen.current = open;
  }, [open]);

  const toggleDropdown = (): void => {
    setOpen((prevOpen) => {
      return !prevOpen;
    });
  };

  const handleClose = (
    event: MouseEvent | TouchEvent | React.MouseEvent<HTMLLIElement> | React.MouseEvent<Document>
  ): void => {
    if (anchorRef.current && anchorRef.current.contains(event.target as Node)) {
      return;
    }
    setOpen(false);
  };

  function handleListKeyDown(event: React.KeyboardEvent): void {
    if (event.key === "Tab") {
      event.preventDefault();
      setOpen(false);
    }
  }

  const DefaultButton = ({
    children,
    ...props
  }: {
    children: ReactNode;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    [key: string]: any;
  }): ReactElement<any> => (
    <UnStyledButton isTextBold isSmallerText {...props}>
      {children}
    </UnStyledButton>
  );

  const Wrapper = props.wrapper ?? DefaultButton;

  const DropdownMenu = (): ReactElement<any> => {
    return (
      <MenuList
        autoFocusItem={open}
        className={`${props.horizontal ? "flex flex-row" : ""}`}
        id="menu-list-grow"
        onKeyDown={handleListKeyDown}
        // eslint-disable-next-line jsx-a11y/aria-role
        role={`${props.horizontal ? "menubar" : "menu"}`}
      >
        {props.dropdownOptions.map((option, i) => {
          return (
            <MenuItem onClick={option.onClick} key={i} className={props.dropdownClassName}>
              <Wrapper {...option.props}>{option.text}</Wrapper>
            </MenuItem>
          );
        })}
      </MenuList>
    );
  };

  return (
    <>
      <UnStyledButton
        onClick={toggleDropdown}
        ref={anchorRef}
        dataTestid={`primary${`-${props.name}`}-dropdown-button`}
      >
        <div className="display-flex flex-row height-full">
          <div
            className={`${
              props.hideDivider ? "" : "padding-right-1"
            } padding-y-1 display-flex flex-justify-center width-100`}
          >
            <div className="flex-align-self-center">{props.children}</div>
          </div>
          <Icon
            className={`usa-icon--size-5 ${
              props.hideDivider ? "" : "border-left-1px "
            } padding-y-1 flex-align-self-center`}
            iconName={open ? "arrow_drop_up" : "arrow_drop_down"}
          />
        </div>
      </UnStyledButton>
      <Popper
        open={open}
        anchorEl={anchorRef.current}
        className="z-100"
        role={undefined}
        transition
        disablePortal={true}
        placement={`${props.horizontal ? "bottom" : "bottom-end"}`}
      >
        {({ TransitionProps, placement }): JSX.Element => {
          return (
            <Grow
              {...TransitionProps}
              style={{
                transformOrigin: placement === "bottom" ? "center top" : "center bottom",
              }}
            >
              <Box className="drop-shadow-xl">
                <Paper>
                  <ClickAwayListener onClickAway={handleClose}>{DropdownMenu()}</ClickAwayListener>{" "}
                </Paper>
              </Box>
            </Grow>
          );
        }}
      </Popper>
    </>
  );
};
