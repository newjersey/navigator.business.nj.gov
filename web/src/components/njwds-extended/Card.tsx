import Link from "next/link";
import React, { ReactElement } from "react";

interface Props {
  headerText: string;
  mainText: string;
  buttonText: string;
  link: string;
  imageSource: string;
  altImageText: string;
  buttonClassName?: string;
  mainTextClassName?: string;
  // All other props
  [x: string]: unknown;
}

export const Card = (props: Props): ReactElement => {
  const {
    headerText,
    mainText,
    buttonText,
    link,
    imageSource,
    altImageText,
    buttonClassName,
    mainTextClassName,
  } = props;

  return (
    <li className="usa-card usa-card--header-first tablet:grid-col">
      <div className="usa-card__container flex-align-center card-container-nj-style">
        <div className="usa-card__header padding-top-8 padding-top-1">
          <h3 className="usa-card__heading text-green text-center ">{headerText}</h3>
        </div>
        <div className={`usa-card__body text-center card-padding no-padding-x ${mainTextClassName || ""}`}>
          <p>{mainText}</p>
        </div>

        <div className="usa-card__footer card-footer-padding">
          <Link href={link} passHref>
            <button className={`usa-button usa-button--outline ${buttonClassName || ""}`}>
              {buttonText}
            </button>
          </Link>
        </div>

        <div className="usa-card__media usa-card__media--inset card-media-padding">
          <div className="usa-card__img white-background img-width-250">
            <img src={imageSource} alt={altImageText} />
          </div>
        </div>
      </div>
    </li>
  );
};
