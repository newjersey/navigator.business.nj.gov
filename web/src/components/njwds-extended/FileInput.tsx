import { flipObject, templateEval } from "@/lib/utils/helpers";
import { AcceptedFileType, InputFile } from "@businessnjgovnavigator/shared";
import {
  ReactElement,
  useEffect,
  useLayoutEffect,
  useRef,
  useState,
} from "react";

interface Props {
  acceptedFileTypes?: {
    errorMessage: string;
    fileTypes: AcceptedFileType[];
  };
  maxFileSize?: {
    errorMessage: string;
    maxSizeInMegabytes: number;
  };
  errorMessageRequired: string;
  hasError?: boolean;
  helperText: string;
  value: InputFile | undefined;
  onChange: (file: InputFile) => void;
}

const FileTypeToInputString: Record<AcceptedFileType, string> = {
  PDF: "application/pdf",
  PNG: "image/png",
};

const InputStringToFileType = flipObject(FileTypeToInputString) as Record<
  string,
  AcceptedFileType
>;

const BYTES_IN_A_MB = 1048576;

const base64ToFile = (params: {
  base64Contents: string;
  type: string;
  filename: string;
}): File => {
  const imageContent = atob(params.base64Contents);
  const buffer = new ArrayBuffer(imageContent.length);
  const view = new Uint8Array(buffer);

  for (let n = 0; n < imageContent.length; n++) {
    const charCode = imageContent.codePointAt(n);
    if (charCode) {
      view[n] = charCode;
    }
  }

  const blob = new Blob([buffer], { type: params.type });
  return new File([blob], params.filename, { type: params.type });
};

export const FileInput = ({
  acceptedFileTypes,
  errorMessageRequired,
  hasError,
  helperText,
  maxFileSize,
  value,
  onChange,
}: Props): ReactElement => {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const fileInput = require("../../../../node_modules/@uswds/uswds/packages/usa-file-input/src");
  const fileInputRef = useRef<HTMLInputElement>(null);
  const size = maxFileSize?.maxSizeInMegabytes
    ? maxFileSize.maxSizeInMegabytes * BYTES_IN_A_MB
    : undefined;

  const [hasFileBeenRead, setHasFileBeenRead] = useState<boolean>(false);
  const [fileUploadError, setFileUploadError] = useState({
    hasError: false,
    errorMessage: "",
  });

  const createAcceptedFileString = (): string | undefined => {
    if (!acceptedFileTypes) return undefined;
    return acceptedFileTypes.fileTypes
      .map((it) => FileTypeToInputString[it])
      .join(",");
  };

  const getErrorMessage = (): string => {
    if (fileUploadError.hasError) {
      return fileUploadError.errorMessage;
    } else if (hasError) {
      return errorMessageRequired;
    }
    return "";
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    if (e.target.files && e.target.files.length > 0) {
      const uploadedFile = e.target.files[0];
      const reader = new FileReader();

      // eslint-disable-next-line unicorn/prefer-add-event-listener
      reader.onload = function (event: ProgressEvent<FileReader>): void {
        if (!event.target?.result) {
          return;
        }

        setHasFileBeenRead(true);
        onChange({
          base64Contents: (event.target.result as string).split(",")[1],
          sizeInBytes: uploadedFile.size,
          fileType: InputStringToFileType[uploadedFile.type],
          filename: uploadedFile.name,
        });
        setFileUploadError({ hasError: false, errorMessage: "" });
      };
      reader.readAsDataURL(uploadedFile);
    }
  };

  useLayoutEffect(() => {
    const fileInputElement = fileInputRef.current;

    const preventInvalidFileSelection = (e: Event): void => {
      const eventTarget = e.target as HTMLInputElement;
      if (eventTarget.files && eventTarget.files.length > 0) {
        const selectedFile = eventTarget.files[0];
        if (
          acceptedFileTypes &&
          !acceptedFileTypes.fileTypes.includes(
            InputStringToFileType[selectedFile.type]
          )
        ) {
          e.stopImmediatePropagation();
          setFileUploadError({
            hasError: true,
            errorMessage: templateEval(acceptedFileTypes.errorMessage, {
              fileName: selectedFile.name,
            }),
          });
        } else if (maxFileSize && size && selectedFile.size > size) {
          e.stopImmediatePropagation();
          setFileUploadError({
            hasError: true,
            errorMessage: templateEval(maxFileSize.errorMessage, {
              fileName: selectedFile.name,
              maxFileSize: `${maxFileSize.maxSizeInMegabytes}`,
            }),
          });
        }
      }
    };

    // Event listener is added to the file input element ahead of USWDS listeners - React onChange will be added after
    fileInputElement?.addEventListener("change", (e: Event) => {
      preventInvalidFileSelection(e);
    });
    if (
      typeof fileInput.on === "function" &&
      typeof fileInput.off === "function"
    ) {
      fileInput.on(fileInputElement); // initialize USWDS fileInput component
    }
  }, []); // eslint-disable-line react-hooks/exhaustive-deps

  const addPreviouslyUploadedFile = (): void => {
    const fileInputElement = fileInputRef.current;
    if (value && fileInputElement?.files?.length === 0) {
      const file = base64ToFile({
        base64Contents: value.base64Contents,
        filename: value.filename,
        type: FileTypeToInputString[value.fileType],
      });

      const dT = new DataTransfer();
      if (fileInputElement) {
        dT.items.add(file);
        fileInputElement.files = dT.files;
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        const event = new Event("change", { target: { files: [file] } });
        fileInputElement.dispatchEvent(event);
      }
    }
  };

  useEffect(() => {
    addPreviouslyUploadedFile();
  }, [value]); // eslint-disable-line react-hooks/exhaustive-deps

  return (
    <>
      <div
        className={`usa-form-group${
          hasError || (fileUploadError.hasError && " usa-form-group--error")
        }`}
      >
        <label className="usa-label" htmlFor="file-input-single">
          {helperText}
        </label>
        <span className="usa-error-message" id="file-input-error-alert">
          {getErrorMessage()}
        </span>
        <input
          data-testid="file-input"
          accept={createAcceptedFileString()}
          className="usa-file-input"
          id="file-input-single"
          name="file-input-single"
          onChange={handleChange}
          size={size}
          ref={fileInputRef}
          type="file"
        />
      </div>
      <div
        aria-hidden="true"
        data-testid={hasFileBeenRead ? "file-is-read" : ""}
      />
    </>
  );
};
