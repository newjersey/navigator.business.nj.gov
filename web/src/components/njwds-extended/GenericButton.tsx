import { CircularProgress } from "@mui/material";
import React, { forwardRef, ReactElement, Ref, useEffect, useRef, useState } from "react";

export interface GenericButtonProps {
  children?: React.ReactNode;
  id?: string;
  isAriaControls?: string;
  isAriaExpanded?: boolean;
  isAriaHaspopup?: boolean;

  onClick?: (() => void) | ((event: React.MouseEvent) => Promise<void>) | ((event: React.MouseEvent) => void);
  dataTestId?: string;
  className?: string;
  isIntercomEnabled?: boolean;
  isSubmitButton?: boolean;

  isLoading?: boolean;
  isRightMarginRemoved?: boolean;
  isNotFullWidthOnMobile?: boolean;
  isFullWidthOnDesktop?: boolean;
  isVerticalPaddingRemoved?: boolean;
  isReducedPaddingRight?: boolean;
}

export const GenericButton = forwardRef(function GenericButton(
  props: GenericButtonProps,
  ref?: Ref<HTMLButtonElement> | undefined
): ReactElement {
  const disabledClass = "usa-button--disabled";
  const showDisabledClass = props.isLoading ? disabledClass : "";
  const noRightMargin = props.isRightMarginRemoved ? "margin-right-0" : "margin-right-2";
  const fullWidth = props.isFullWidthOnDesktop ? "width-100 flex fjc" : "";
  const isNotFullWidthOnMobile = props.isNotFullWidthOnMobile ? "width-auto" : "";
  const isVerticalPaddingRemoved = props.isVerticalPaddingRemoved ? "padding-y-0" : "padding-y-11px";
  const intercomButton = props.isIntercomEnabled ? "intercom-button" : "";
  const isReducedPaddingRight = props.isReducedPaddingRight ? "padding-right-1" : "";

  const widthRef = useRef<HTMLInputElement | null>(null);
  const [width, setWidth] = useState<number>();
  const [height, setHeight] = useState<number>();
  useEffect(() => {
    if (!props.isLoading && widthRef.current !== null) {
      setWidth(widthRef.current.clientWidth);
      setHeight(widthRef.current.clientHeight);
    }
  }, [height, width, props.isLoading, widthRef]);

  const className = [
    props.className,
    isNotFullWidthOnMobile,
    noRightMargin,
    intercomButton,
    showDisabledClass,
    fullWidth,
    isVerticalPaddingRemoved,
    isReducedPaddingRight,
  ]
    .map((i) => {
      return i?.trim();
    })
    .filter((value: string | undefined) => {
      return value && value.length > 0;
    })
    .join(" ");

  return (
    <button
      className={className}
      onClick={props.onClick}
      ref={ref}
      {...(props.isSubmitButton ? { type: "submit" } : { type: "button" })}
      {...(props.dataTestId ? { "data-testid": props.dataTestId } : {})}
      {...(props.id ? { id: props.id } : {})}
      {...(props.isAriaControls ? { "aria-controls": props.isAriaControls } : {})}
      {...(props.isAriaExpanded ? { "aria-expanded": props.isAriaExpanded } : {})}
      {...(props.isAriaHaspopup ? { "aria-haspopup": props.isAriaHaspopup } : {})}
    >
      {props.isLoading ? (
        <div style={{ width: width, height: height }} data-testid="loading-spinner">
          <div className="padding-top-1px">
            <CircularProgress size={24} thickness={10} />
          </div>
        </div>
      ) : (
        <div ref={widthRef} className={`display-flex flex-row flex-align-center flex-justify-center`}>
          {props.children}
        </div>
      )}
    </button>
  );
});
