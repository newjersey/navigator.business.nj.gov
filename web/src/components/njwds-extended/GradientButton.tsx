import { Icon } from "@/components/njwds/Icon";
import { ReactElement } from "react";

interface Props {
  text: string;
  ariaLabel?: string;
  dataTestid?: string;
  icon?: string;
  onClick?: () => void;
  role?: string;
}

export const GradientButton = (props: Props): ReactElement => {
  return (
    <button
      {...(props.ariaLabel ? { "aria-label": props.ariaLabel } : {})}
      {...(props.dataTestid ? { "data-testid": props.dataTestid } : {})}
      {...(props.role ? { role: props.role } : {})}
      className="gradient-button"
      onClick={props.onClick}
    >
      {props.icon && (
        <img src={`/img/${props.icon}.svg`} alt={`${props.icon} icon`} role="presentation" />
      )}
      <div className="gradient-button-text">{props.text}</div>
      <Icon className="usa-icon--size-4 text-base" iconName="navigate_next" />
    </button>
  );
};
