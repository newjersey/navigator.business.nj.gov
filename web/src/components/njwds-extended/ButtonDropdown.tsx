import { Button } from "@/components/njwds-extended/Button";
import { Icon } from "@/components/njwds/Icon";
import { ClickAwayListener, Grow, MenuItem, MenuList, Paper, Popper } from "@mui/material";
import React, { ReactElement, ReactNode, useEffect, useRef, useState } from "react";

type ButtonDropdownOption = {
  text: string;
  onClick: () => void;
};

interface Props {
  children: ReactNode;
  dropdownOptions: ButtonDropdownOption[];
}

export const ButtonDropdown = (props: Props): ReactElement => {
  const [open, setOpen] = useState(false);
  const anchorRef = useRef<HTMLButtonElement | null>(null);
  const prevOpen = useRef(open);
  useEffect(() => {
    if (null !== anchorRef.current && prevOpen.current && !open) {
      anchorRef.current.focus();
    }
    prevOpen.current = open;
  }, [open]);

  const toggleDropdown = () => {
    setOpen((prevOpen) => !prevOpen);
  };

  const handleClose = (
    event: MouseEvent | TouchEvent | React.MouseEvent<HTMLLIElement> | React.MouseEvent<Document>
  ): void => {
    if (anchorRef.current && anchorRef.current.contains(event.target as Node)) {
      return;
    }
    setOpen(false);
  };

  function handleListKeyDown(event: React.KeyboardEvent) {
    if (event.key === "Tab") {
      event.preventDefault();
      setOpen(false);
    }
  }

  const DropdownMenu = () => (
    <MenuList autoFocusItem={open} id="menu-list-grow" onKeyDown={handleListKeyDown}>
      {props.dropdownOptions.map((option, i) => (
        <MenuItem onClick={option.onClick} key={i}>
          <Button style="tertiary" textBold smallText>
            {option.text}
          </Button>
        </MenuItem>
      ))}
    </MenuList>
  );

  return (
    <>
      <button
        className="usa-button padding-y-0 padding-right-0"
        onClick={toggleDropdown}
        type="button"
        ref={anchorRef}
      >
        <div className="display-flex flex-row height-full">
          <div className="padding-right-2 padding-y-1 display-flex flex-justify-center width-100">
            <div className="flex-align-self-center">{props.children}</div>
          </div>
          <Icon className="usa-icon--size-5 border-left-1px padding-y-1 flex-align-self-center">
            arrow_drop_down
          </Icon>
        </div>
      </button>

      <Popper open={open} anchorEl={anchorRef.current} role={undefined} transition disablePortal={true}>
        {({ TransitionProps, placement }) => (
          <Grow
            {...TransitionProps}
            style={{
              transformOrigin: placement === "bottom" ? "center top" : "center bottom",
            }}
          >
            <Paper>
              <ClickAwayListener onClickAway={handleClose}>{DropdownMenu()}</ClickAwayListener>
            </Paper>
          </Grow>
        )}
      </Popper>
    </>
  );
};
