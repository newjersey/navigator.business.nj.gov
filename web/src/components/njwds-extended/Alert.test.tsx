import { Alert } from "@/components/njwds-extended/Alert";
import { render } from "@testing-library/react";

describe("Alert", () => {
  describe("Variants", () => {
    it("Error displays correctly", () => {
      const view = render(<Alert variant="error">Test Button</Alert>).baseElement;
      expect(view).toMatchSnapshot();
    });

    it("Info displays correctly", () => {
      const view = render(<Alert variant="info">Test Button</Alert>).baseElement;
      expect(view).toMatchSnapshot();
    });

    it("Success displays correctly", () => {
      const view = render(<Alert variant="success">Test Button</Alert>).baseElement;
      expect(view).toMatchSnapshot();
    });

    it("warning displays correctly", () => {
      const view = render(<Alert variant="warning">Test Button</Alert>).baseElement;
      expect(view).toMatchSnapshot();
    });
  });

  it("warning with heading displays correctly", () => {
    const view = render(
      <Alert variant="warning" heading="Warning Alert with heading">
        Test Button
      </Alert>
    ).baseElement;
    expect(view).toMatchSnapshot();
    expect(view).toHaveTextContent("Warning Alert with heading");
  });

  it("error with no icon displays correctly", () => {
    const view = render(
      <Alert variant="error" noIcon>
        Test Button
      </Alert>
    ).baseElement;
    expect(view).toMatchSnapshot();
  });

  it("info with rounded corners displays correctly", () => {
    const view = render(
      <Alert variant="info" rounded>
        Test Button
      </Alert>
    ).baseElement;
    expect(view).toMatchSnapshot();
  });
});
