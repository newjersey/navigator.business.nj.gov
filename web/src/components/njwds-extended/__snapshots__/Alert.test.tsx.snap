// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert Variants Error displays correctly 1`] = `
<body>
  <div>
    <div
      class="usa-alert margin-y-2 usa-alert--slim usa-alert--error"
      role="alert"
    >
      <div
        class="usa-alert__body"
      >
        <div
          class="usa-alert__text"
        >
          Test Button
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Alert Variants Info displays correctly 1`] = `
<body>
  <div>
    <div
      class="usa-alert margin-y-2 usa-alert--slim usa-alert--info"
    >
      <div
        class="usa-alert__body"
      >
        <div
          class="usa-alert__text"
        >
          Test Button
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Alert Variants Success displays correctly 1`] = `
<body>
  <div>
    <div
      class="usa-alert margin-y-2 usa-alert--slim usa-alert--success"
    >
      <div
        class="usa-alert__body"
      >
        <div
          class="usa-alert__text"
        >
          Test Button
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Alert Variants warning displays correctly 1`] = `
<body>
  <div>
    <div
      class="usa-alert margin-y-2 usa-alert--slim usa-alert--warning"
    >
      <div
        class="usa-alert__body"
      >
        <div
          class="usa-alert__text"
        >
          Test Button
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Alert error with no icon displays correctly 1`] = `
<body>
  <div>
    <div
      class="usa-alert margin-y-2 usa-alert--slim usa-alert--error usa-alert--no-icon"
      role="alert"
    >
      <div
        class="usa-alert__body"
      >
        <div
          class="usa-alert__text"
        >
          Test Button
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Alert info with rounded corners displays correctly 1`] = `
<body>
  <div>
    <div
      class="usa-alert margin-y-2 usa-alert--slim radius-md usa-alert--info"
    >
      <div
        class="usa-alert__body"
      >
        <div
          class="usa-alert__text"
        >
          Test Button
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Alert warning with heading displays correctly 1`] = `
<body>
  <div>
    <div
      class="usa-alert margin-y-2 usa-alert--slim usa-alert--warning"
    >
      <div
        class="usa-alert__body"
      >
        <h3
          class="margin-bottom-0"
        >
          Warning Alert with heading
        </h3>
        <div
          class="usa-alert__text"
        >
          Test Button
        </div>
      </div>
    </div>
  </div>
</body>
`;
